File: packages/sdk/hush-go/merkle.go
package hush

import (
	"errors"
	"runtime"
)

// MerkleRoot computes the Merkle root from a JSON array of hex-encoded leaf hashes.
// Returns the root hash as a hex string.
func MerkleRoot(leafHashesJSON string) (string, error) {
	runtime.LockOSThread()
	defer runtime.UnlockOSThread()

	cl, err := allocCString(leafHashesJSON)
	if err != nil {
		return "", err
	}
	defer freeCString(cl)

	p := ffiMerkleRoot(cl)
	if p == nil {
		return "", lastError()
	}
	return goStringFromCFree(p), nil
}

// MerkleProof generates a Merkle inclusion proof for the leaf at the given index.
// Returns the proof as a JSON string.
func MerkleProof(leafHashesJSON string, index int) (string, error) {
	if index < 0 {
		return "", errors.New("hush: index must be non-negative")
	}

	runtime.LockOSThread()
	defer runtime.UnlockOSThread()

	cl, err := allocCString(leafHashesJSON)
	if err != nil {
		return "", err
	}
	defer freeCString(cl)

	p := ffiMerkleProof(cl, index)
	if p == nil {
		return "", lastError()
	}
	return goStringFromCFree(p), nil
}

// VerifyMerkleProof verifies a Merkle inclusion proof.
// Returns true if the proof is valid.
func VerifyMerkleProof(leafHex, proofJSON, rootHex string) (bool, error) {
	runtime.LockOSThread()
	defer runtime.UnlockOSThread()

	clh, err := allocCString(leafHex)
	if err != nil {
		return false, err
	}
	defer freeCString(clh)
	cpj, err := allocCString(proofJSON)
	if err != nil {
		return false, err
	}
	defer freeCString(cpj)
	crh, err := allocCString(rootHex)
	if err != nil {
		return false, err
	}
	defer freeCString(crh)

	rc := ffiVerifyMerkleProof(clh, cpj, crh)
	switch rc {
	case 1:
		return true, nil
	case 0:
		return false, nil
	default:
		return false, lastError()
	}
}

File: packages/sdk/hush-csharp/src/Hush/Merkle/Merkle.cs
using System;
using Hush.Internal;

namespace Hush.Merkle
{
    /// <summary>
    /// Merkle tree operations: root computation, proof generation, and verification.
    /// </summary>
    public static class Merkle
    {
        /// <summary>
        /// Compute the Merkle root from a JSON array of hex-encoded leaf hashes.
        /// </summary>
        /// <param name="leafHashesJson">JSON array of hex strings, e.g. <c>["aabb..","ccdd.."]</c>.</param>
        /// <returns>The hex-encoded Merkle root.</returns>
        public static string ComputeRoot(string leafHashesJson)
        {
            if (leafHashesJson == null) throw new ArgumentNullException(nameof(leafHashesJson));
            Utf8Validation.RejectEmbeddedNul(leafHashesJson, nameof(leafHashesJson));
            var ptr = NativeMethods.hush_merkle_root(leafHashesJson);
            HushException.ThrowIfNull(ptr);
            using var ns = new NativeString(ptr);
            return ns.ToString()!;
        }

        /// <summary>
        /// Generate a Merkle inclusion proof for the leaf at the given index.
        /// </summary>
        /// <param name="leafHashesJson">JSON array of hex-encoded leaf hashes.</param>
        /// <param name="index">Zero-based leaf index.</param>
        /// <returns>The proof as a JSON string.</returns>
        public static string GenerateProof(string leafHashesJson, int index)
        {
            if (leafHashesJson == null) throw new ArgumentNullException(nameof(leafHashesJson));
            Utf8Validation.RejectEmbeddedNul(leafHashesJson, nameof(leafHashesJson));
            if (index < 0)
                throw new ArgumentOutOfRangeException(nameof(index), "Index must be non-negative.");
            var ptr = NativeMethods.hush_merkle_proof(leafHashesJson, (UIntPtr)index);
            HushException.ThrowIfNull(ptr);
            using var ns = new NativeString(ptr);
            return ns.ToString()!;
        }

        /// <summary>
        /// Verify a Merkle inclusion proof.
        /// </summary>
        /// <param name="leafHex">Hex-encoded leaf hash.</param>
        /// <param name="proofJson">The proof as a JSON string (from <see cref="GenerateProof"/>).</param>
        /// <param name="rootHex">The expected hex-encoded Merkle root.</param>
        /// <returns>True if the proof is valid.</returns>
        public static bool VerifyProof(string leafHex, string proofJson, string rootHex)
        {
            if (leafHex == null) throw new ArgumentNullException(nameof(leafHex));
            if (proofJson == null) throw new ArgumentNullException(nameof(proofJson));
            if (rootHex == null) throw new ArgumentNullException(nameof(rootHex));
            Utf8Validation.RejectEmbeddedNul(leafHex, nameof(leafHex));
            Utf8Validation.RejectEmbeddedNul(proofJson, nameof(proofJson));
            Utf8Validation.RejectEmbeddedNul(rootHex, nameof(rootHex));

            var rc = NativeMethods.hush_verify_merkle_proof(leafHex, proofJson, rootHex);
            if (rc < 0)
                throw HushException.FromLastError();
            return rc == 1;
        }
    }
}
